File: 10 - Queues/java-files/Circular_queue.java

// CIRCULAR QUEUE
// TC : O(1)

public class Circular_queue{
    int size;
    int front;
    int rear;
    int arr[];

    Circular_queue(int size){
        this.size=size;
        this.front=-1;
        this.rear=-1;
        arr=new int[size];
    }

    /* ADDING ELEMENT ( PUSH OPERATION )
    1.Check whether queue is Full – Check ((rear == SIZE-1 && front == 0) || (rear == front-1)).
    2.If it is full then display Queue is full. If queue is not full then, check if (rear == SIZE – 1 && front != 0) if it is true then set rear=0 and insert element. */
    void enqueue(int data){
        if(front==0 && rear==size-1 || rear==(front-1)%(size-1)){
            // QUEUE IS FULL
            System.out.println("QUEUE IS FULL");
        }
        else if(front==-1){
            // FIRST ELEMENT
            front=rear=0;
        }
        else if(rear==size-1 && front!=0){
            // REAR KO AAGE LANE KE LIYE
            rear=0;
        }
        else{ // NORMAL FLOW
            rear++;
        }
        // PUSHING
        arr[rear]=data;
    }

    /*  REMOVE ELEMENT ( POP OPERATION )
    1.Check whether queue is Empty means check (front==-1).
    2.If it is empty then display Queue is empty. If queue is not empty then step 3
    3.Check if (front==rear) if it is true then set front=rear= -1 else check if (front==size-1), if it is true then set front=0 and return the element.*/
    int dequeue(){
        if(front==-1){
            // QUEUE IS EMPTY
            System.out.println("QUEUE IS EMPTY");
            return -1;
        }
        int ans=arr[front];
        arr[front]=-1;
        if(front==rear){
            // SINGLE ELEMENT
            front=rear=-1;
        }
        else if(front==size-1){
            // TO MAINTAIN CYCLIC NATURE
            front=0;
        }
        else{
            // NORMAL FLOW
            front++;
        }
        return ans;
    }

    public static void main(String[] args) {
        Circular_queue q=new Circular_queue(8);
        q.enqueue(67);
        q.enqueue(55);
        System.out.println(q.dequeue());
        System.out.println(q.dequeue()); 
    }
}

File: 10 - Queues/circularQueue.cpp
#include <iostream>
using namespace std;

// % can be used in circular queues but should avoid
// range of % 0 to size -1.

class Cqueue{

    public:
    int *arr;
    int size;
    int front;
    int rear;

    Cqueue(int size){
        arr=new int[size];
        this->size=size;
        front=-1;
        rear=-1;
    }

    void push(int val){
        // overflow
        // special case important
        if(front==0 && rear==size-1 || rear==front-1){
            cout << "Overflow" << endl;
        }

        // empty case -> first element
        // important case
        else if(front==-1 && rear==-1){
            front++;
            rear++;
            arr[rear]=val;
        }
        // circular nature
        else if(rear==size-1 && front!=0){
            rear=0;
            arr[rear]=val;
        }
        // normal flow
        else{
            rear++;
            arr[rear]=val;
        }
    }

    void pop(){
        // underflow
        if(front==-1 && rear==-1){
            cout << "Undeflow" << endl;
        }
        // 1 element
        // important case
        else if(front==rear){
            arr[front]=-1;
            front=-1;
            rear=-1;
        }
        // circular
        else if(front==size-1){
            arr[front]=-1;
            front =0;
        }
        // normal flow
        else{
            arr[front]=-1;
            front++;
        }
    }

    void print(){
        for(int i=0;i<size;i++){
            cout << arr[i]<<" ";
        }
        cout << endl;
    }
};

int main(){

    Cqueue q(5);
    q.print();

    q.push(10);
    q.print();

    q.push(20);
    q.print();

    q.push(30);
    q.print();

    q.push(40);
    q.print();

    q.push(50);
    q.print();

    q.pop();
    q.pop();
    q.print();

    q.push(60);
    q.print();

    q.push(70);
    q.print();

    q.push(80);
    q.print();
}